File: sdks/sandbox/javascript/src/adapters/ptyAdapter.ts
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { ExecdClient } from "../openapi/execdClient.js";
import { throwOnOpenApiFetchError } from "./openapiError.js";
import type { PtySession, PtySessionStatus } from "../models/execd.js";
import type { ExecdPty } from "../services/execdPty.js";
import { SandboxError, SandboxException } from "../core/exceptions.js";

export class PtyAdapter implements ExecdPty {
  constructor(private readonly client: ExecdClient) {}

  async createSession(opts?: { cwd?: string; command?: string }): Promise<PtySession> {
    const { data, error, response } = await this.client.POST("/pty", {
      body: { cwd: opts?.cwd, command: opts?.command },
    });
    throwOnOpenApiFetchError({ error, response }, "Create PTY session failed");
    return { sessionId: data!.session_id };
  }

  async getSession(sessionId: string): Promise<PtySessionStatus> {
    const { data, error, response } = await this.client.GET("/pty/{sessionId}", {
      params: { path: { sessionId } },
    });
    throwOnOpenApiFetchError({ error, response }, "Get PTY session failed");
    return {
      sessionId: data!.session_id,
      running: data!.running,
      outputOffset: data!.output_offset,
    };
  }

  async deleteSession(sessionId: string): Promise<void> {
    // Success is an empty 200 body (Content-Length: 0), which openapi-fetch skips
    // parsing; error responses (e.g. 404 CONTEXT_NOT_FOUND, 501 NOT_SUPPORTED) keep
    // their JSON code/message so throwOnOpenApiFetchError can surface them.
    const { error, response } = await this.client.DELETE("/pty/{sessionId}", {
      params: { path: { sessionId } },
    });
    throwOnOpenApiFetchError({ error, response }, "Delete PTY session failed");
  }
}

/**
 * Fallback PTY service used when a custom {@link AdapterFactory} does not supply a
 * PTY adapter. Keeps `sandbox.pty` defined while failing loudly on use, so the
 * execd stack contract stays additive for pre-existing factories.
 */
export class UnavailablePtyAdapter implements ExecdPty {
  private failure(): SandboxException {
    return new SandboxException({
      message:
        "PTY service is not available: the configured adapter factory did not provide a PTY adapter.",
      error: new SandboxError(SandboxError.INVALID_ARGUMENT, "PTY service unavailable"),
    });
  }

  createSession(): Promise<PtySession> {
    return Promise.reject(this.failure());
  }

  getSession(): Promise<PtySessionStatus> {
    return Promise.reject(this.failure());
  }

  deleteSession(): Promise<void> {
    return Promise.reject(this.failure());
  }
}

File: sdks/sandbox/csharp/src/OpenSandbox/Adapters/PtyAdapter.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Text.Json.Serialization;
using OpenSandbox.Internal;
using OpenSandbox.Models;
using OpenSandbox.Services;

namespace OpenSandbox.Adapters;

/// <summary>
/// Adapter for the execd interactive PTY session service.
/// </summary>
internal sealed class PtyAdapter : IExecdPty
{
    private readonly HttpClientWrapper _client;

    public PtyAdapter(HttpClientWrapper client)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
    }

    public Task<PtySession> CreateSessionAsync(
        string? cwd = null,
        string? command = null,
        CancellationToken cancellationToken = default)
    {
        var body = new PtyCreateRequest { Cwd = cwd, Command = command };
        return _client.PostAsync<PtySession>("/pty", body, cancellationToken);
    }

    public Task<PtySessionStatus> GetSessionAsync(string sessionId, CancellationToken cancellationToken = default)
    {
        return _client.GetAsync<PtySessionStatus>(
            $"/pty/{Uri.EscapeDataString(sessionId)}",
            cancellationToken: cancellationToken);
    }

    public Task DeleteSessionAsync(string sessionId, CancellationToken cancellationToken = default)
    {
        return _client.DeleteAsync(
            $"/pty/{Uri.EscapeDataString(sessionId)}",
            cancellationToken: cancellationToken);
    }

    private sealed class PtyCreateRequest
    {
        [JsonPropertyName("cwd")]
        public string? Cwd { get; set; }

        [JsonPropertyName("command")]
        public string? Command { get; set; }
    }
}
